diff --git a/torch_scatter/scatter.py b/torch_scatter/scatter.py
index b994873..00617b7 100644
--- a/torch_scatter/scatter.py
+++ b/torch_scatter/scatter.py
@@ -1,6 +1,7 @@
 from typing import Optional, Tuple
 
 import torch
+import mx_driving
 
 from .utils import broadcast
 
@@ -157,6 +158,6 @@ def scatter(src: torch.Tensor, index: torch.Tensor, dim: int = -1,
     elif reduce == 'min':
         return scatter_min(src, index, dim, out, dim_size)[0]
     elif reduce == 'max':
-        return scatter_max(src, index, dim, out, dim_size)[0]
+        return mx_driving.scatter_max(src.float(),index.to(torch.int32),out)[0]
     else:
         raise ValueError
